### ISSUE TITLE
[BUG] Cython incorrectly searches for template types inside templates

### ISSUE THREAD

**Issue by maxbachmann** [cython/cython#4546]
**Describe the bug**
When returning a template class from a template function without setting the template type, Cython matches this template type incorrectly.

**To Reproduce**
The following code does not set the template type of the Test class. However it compiles successfully, since both `test_func` and `Test` name the template parameter `T`.
```cython
cdef extern from "*":
    cdef cppclass Test[T]:
        pass

    cdef Test test_func[T]()

cdef test():
    cdef Test[double] result = test_func[double]()
```
When naming the template types differently:
```cython
cdef extern from "*":
    cdef cppclass Test[U]:
        pass

    cdef Test test_func[T]()

cdef test():
    cdef Test[double] result = test_func[double]()
```
this fails to compile with the following error:
```
Cannot assign type 'Test[U]' to 'Test[double]'
```

**Expected behavior**
Cython should not match template types of different classes. So the example shown above should not compile successfully.

**Environment (please complete the following information):**
 - OS: Linux
 - Python version 3.9
 - Cython version 3.0.0a9

**Comment by da-woods** [comment]
```
cdef Test test_func[T]()
```

I think you'd have to do

```
template <typename T>
Test<T> test_func<T>()
```

In c++ (i.e. explicitly specify the return type). So I'm not sure that Cython should be different. I could be missing the point though

**Comment by maxbachmann** [comment]
> In c++ (i.e. explicitly specify the return type). So I'm not sure that Cython should be different. I could be missing the point though.

Yes I just expect Cython to raise an error when I forget to specify the template type ;)
